[FLINK-23123] Add at-least-once Kafka Sink based on FLIP-143

fapaul opened the pull request:
A few things are still missing

- [ ] Forward Kafka Metrics
- [x] Handle state initialization i.e. for recovery
- [x] Extend test coverage with induced failures

but I think this is a first good draft to get some feedback specifically for the user-facing interfaces and the public API changes.

gaoyunhaii committed:
[FLINK-21086][runtime][checkpoint] Add EndOfUserRecordsEvent

gaoyunhaii committed:
[FLINK-21086][runtime][checkpoint] Let result partition supports waiting for the downstream tasks to processed all the records

gaoyunhaii committed:
[FLINK-21086][runtime][checkpoint] StreamTask waits for all the records get processed by downstream tasks

gaoyunhaii committed:
[FLINK-21086][runtime][checkpoint] Downstream tasks response with Ack message when received EndOfUserRecordsEvent

EronWright committed:
[FLINK-22700] [api] Propagate watermarks to sink API (FLIP-167)

This closes #15950.

PatrickRen committed:
[FLINK-22766][connector/kafka] Report offsets and Kafka consumer metrics in Flink metric group

KarmaGYZ committed:
[FLINK-21925][core] Introduce public SlotSharingGroup

Users can use this class to describe the name and the the different resource components of a slot sharing group.

lirui-apache committed:
[FLINK-22884][hive] HiveCatalog should mark views as generic and store schema in properties

This closes #16149

Myasuka committed:
[FLINK-22528][docs] Document latency tracking metrics for state accesses

echauchot committed:
[hotfix][release-notes] Update release notes for 1.13 with accumulators semantics change in MiniClusterJobClient

FLINK-18685 changed the semantics of the MiniClusterJobClient. This commit updates the 1.13 release notes
accordingly.

This closes #16256.

fapaul committed:
[FLINK-23123] Add at-least-once Kafka Sink based on FLIP-143

fapaul committed:
[FLINK-23123][connectorst/kafka] Add KafkaProducerRecordSerializationSchema for FLIP-143 Kafka Sink

KarmaGYZ committed:
[FLINK-21925][core] Introduce #slotSharingGroup(SlotSharingGroup) for configuring slot sharing group with its resource

KarmaGYZ committed:
[FLINK-21925][core] Introduce StreamExecutionEnvironment#registerSlotSharingGroup

dawidwys committed:
[FLINK-22972][datastream] Forbid emitting from closed StreamSourceContexts

dawidwys committed:
[FLINK-22972][datastream] Remove StreamOperator#dispose in favour of close and finish

This commit cleans up StreamOperator API in regards to the termination phase and introduces a clean finish() method for flushing all records without releasing resources.

The StreamOperator#close  method which is supposed to flush all records, but at the same time, currently, it closes all resources, including connections to external systems. We need separate methods for flushing and closing resources because we might need the connections when performing the final checkpoint, once all records are flushed. Moreover, the logic for closing resources is duplicated in the StreamOperator#dispose  method.

This closes #16351

gaoyunhaii committed:
[FLINK-21086][runtime][checkpoint] Make CheckpointBarrierHandler supports alignment with finished channels

KarmaGYZ committed:
[FLINK-21925][core] Introduce fine-grained.shuffle-mode.all-blocking to avoid resource deadlock in batch jobs that apply fine-grained resource management

This closes #16307

fapaul committed:
[FLINK-XXXX][core] Expose UserClassLoader and parallel subtasks number to Sink InitContext

luoyuxia committed:
[FLINK-22898][Connectors/Hive] return default-parallelism in HiveParallelismInference when there is no source infer

This closes #16310

PatrickRen committed:
[hotfix][testutil] Add test utilization for listening metric registration

KarmaGYZ committed:
[FLINK-21925][core] Introduce SlotSharingGroupUtils which contains utility for SlotSharingGroup

KarmaGYZ committed:
[hotfix][core] Fix typo in the Builder of ResourceSpec

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/FlinkKafkaSinkContext.java:
```diff
@@ -0,0 +1,44 @@
+
+/** Context providing information of the kafka record target location. */
+@Internal
+public interface FlinkKafkaSinkContext {
```
We should drop the Flink. I hope our users know what they are using.

AHeise replied:
Maybe move into `KafkaProducerRecordSerializationSchema` and rename to `(Serialize)Context`.

fapaul replied:
I am open to renaming it but moving it to the `KafkaProducerRecordSerializationSchema` feels a bit strange because I would expect everything in `KafkaProducerRecordSerializationSchema` needs to be implemented by the user and the context will always come from us.

AHeise replied:
That argument doesn't hold for `KafkaSerializationSchema$InitializationContext`, `SinkWriter$Context`, or `Sink$InitContext/ProcessingTimeService`. I'm open for alternative ways; I'm mostly aiming for consistency.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaProducerRecordSerializationSchema.java:
```diff
@@ -0,0 +1,64 @@
+ * @param <T> the type of values being serialized
+ */
+@PublicEvolving
+public interface KafkaProducerRecordSerializationSchema<T> extends Serializable {
```
`KafkaRecordSerializationSchema` to keep symmetric to `KafkaRecordDeserializationSchema` ?

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaProducerRecordSerializationSchema.java:
```diff
@@ -0,0 +1,64 @@
+     *
+     * @param context Contextual information that can be used during initialization.
+     */
+    default void open(SerializationSchema.InitializationContext context) throws Exception {}
```
I think we should have our own `InitializationContext`. I think this is a mistake in the deserializer.

fapaul replied:
WDYT about the following?
```suggestion
    default void open(DeserializationSchema.InitializationContext context) throws Exception {}
```
Or do you want to introduce a completely new context interface?

AHeise replied:
That change doesn't make any sense :p. 
I'd create a new `InitializationContext` to be more independent.
But tbh I confused `SerializationSchema` with `KafkaSerializationSchema` here, so there may not be a big need.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaProducerRecordSerializationSchema.java:
```diff
@@ -0,0 +1,64 @@
+     * @param timestamp timestamp
+     * @return Kafka {@link ProducerRecord}
+     */
+    ProducerRecord<byte[], byte[]> serialize(
```
Move timestamp into context.

fapaul replied:
The timestamp is already provided by another context [1] that is passed to `SinkWriter#write`. Merging it to the `FlinkKafkaSinkContext` IMO does not really make sense because currently the context is mainly used to let the `KafkaWriter` access information that is only present in the InitContext [2].

[1] https://github.com/apache/flink/blob/488e037285565cc5965c5130ecc92d8a727a4866/flink-core/src/main/java/org/apache/flink/api/connector/sink/SinkWriter.java#L67
[2] https://github.com/apache/flink/blob/488e037285565cc5965c5130ecc92d8a727a4866/flink-core/src/main/java/org/apache/flink/api/connector/sink/Sink.java#L85

AHeise replied:
I don't quite get your arguments. My main idea: you don't need timestamp in 99% of the time (it's usually part of the payload). The old interface didn't even have the parameter. If there is still a need to have that information, the user can access it through `FlinkKafkaSinkContext` that contains all metadata about Kafka and the record.

fapaul replied:
I think the old interface had this parameter [1]. Currently I also do not see a nice way how to pass the timestamp query to the context because it may also be a new context during `write` [2]

[1] https://github.com/fapaul/flink/blob/a6dcb7e41155cb5c209717d0fb7d9c76fa4b61a9/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaSerializationSchema.java#L60
[2] https://github.com/fapaul/flink/blob/a6dcb7e41155cb5c209717d0fb7d9c76fa4b61a9/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaWriter.java#L93

AHeise commented on flink-core/src/main/java/org/apache/flink/api/connector/sink/Sink.java:
```diff
@@ -93,6 +102,9 @@
         int getSubtaskId();
 
+        /** @return number of parallel Sink tasks. */
+        int getNumberOfParallelSubtasks();
```
Yes, we really need that. I'm wondering why noone noticed that before.

AHeise commented on flink-core/src/main/java/org/apache/flink/api/common/serialization/RuntimeContextUserCodeClassLoaderAdapter.java:
```diff
@@ -0,0 +1,42 @@
+/*
```
This is not an hotfix at all. Component should be `core`.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/DefaultFlinkKafkaSinkContext.java:
```diff
@@ -0,0 +1,82 @@
+
+    private final Sink.InitContext context;
+    private final Properties kafkaProducerConfig;
+    private transient FlinkKafkaInternalProducer<?, ?> cachedProducer;
```
Maybe we should use `AdminClient` directly?

fapaul replied:
Is there any benefit in using the `AdminClient`? AFAIK there is no direct method to access the partitions besides describing the topic and manually parsing the partitions. 
The current way is how the old sink has done it.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/DefaultFlinkKafkaSinkContext.java:
```diff
@@ -0,0 +1,82 @@
+                    public int compare(PartitionInfo o1, PartitionInfo o2) {
+                        return Integer.compare(o1.partition(), o2.partition());
+                    }
+                });
```
`Comparator.comparing(PartitionInfo::partition)`

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/DefaultFlinkKafkaSinkContext.java:
```diff
@@ -0,0 +1,82 @@
+        final int[] partitions = new int[partitionsList.size()];
+        for (int i = 0; i < partitions.length; i++) {
+            partitions[i] = partitionsList.get(i).partition();
+        }
```
I think we should use Java stream for this whole method (incl sorting).

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/DefaultFlinkKafkaSinkContext.java:
```diff
@@ -0,0 +1,82 @@
+            partitions[i] = partitionsList.get(i).partition();
+        }
+
+        return partitions;
```
Do we want to add a cache here? If not we should document it on the interface that the user should cache.
I think it would be safest to cache on our side.

fapaul replied:
I've updated to cache the producer externally. Thanks for thint.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/FlinkKafkaInternalProducer.java:
```diff
@@ -0,0 +1,263 @@
+/**
+ * A {@link KafkaProducer} that exposes private fields to allow resume producing from a given state.
+ */
+public class FlinkKafkaInternalProducer<K, V> extends KafkaProducer<K, V> {
```
If we don't need reflection magic anymore, we should remove this class. Please also add @Internal.
Btw can't we just reuse the existing InternalProducer?

fapaul replied:
I fully agree once we get rid of the reflection magic we can drop this class. I made the class package-private, is this enough or is the `@internal` annotation necessary?

The old InternalProducer is very similar **but** I decided to simply extend the `KafkaProducer` rather than implementing the interface and delegating to a `KafkaProducer`.
Prior the delegation was needed due to [1] where every call had to be wrapped.

[1] https://github.com/fapaul/flink/blob/519586b7caa70e7ee900b0ff8affb7536f70e413/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/FlinkKafkaInternalProducer.java#L70

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaSink.java:
```diff
@@ -0,0 +1,197 @@
+                    "Property [{}] not specified. Setting it to {}",
+                    ProducerConfig.TRANSACTION_TIMEOUT_CONFIG,
+                    DEFAULT_KAFKA_TRANSACTION_TIMEOUT);
+        }
```
Move validation to builder ideally directly to setter where applicable. Cross-validation in `build`.

lirui-apache committed:
[FLINK-23168][table-common][table-api-java] GenericInMemoryCatalog shouldn't merge properties for alter DB operation

lirui-apache committed:
[FLINK-23168][hive] HiveCatalog shouldn't merge properties for alter DB operation

This closes #16335

AHeise reviewed (commented):
Some suggestions

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaWriter.java:
```diff
@@ -0,0 +1,203 @@
+    private final int kafkaProducerPoolSize;
+    private final KafkaProducerRecordSerializationSchema<IN> recordSerializer;
+    private final Callback deliveryCallback;
+    private final AtomicLong pendingRecords = new AtomicLong();
+    private final FlinkKafkaSinkContext kafkaSinkContext;
+    private final List<KafkaWriterState> states = new CopyOnWriteArrayList<>();
```
Why do we need sync fields?

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaWriterState.java:
```diff
@@ -0,0 +1,93 @@
+        this(kafkaProducerConfig, producerId, epoch, null);
+    }
+
+    public FlinkKafkaInternalProducer<byte[], byte[]> getProducer() {
+        if (producer == null) {
+            producer = new FlinkKafkaInternalProducer<>(kafkaProducerConfig);
+            producer.resumeTransaction(producerId, epoch);
+        }
+        return producer;
+    }
+
```
Keep state dumb. Just return producerId and let using code constructor the producer.

fapaul replied:
Hmm, I really like the abstraction that consumers only call `getProducer` and do not have to worry about the initialization but I agree a state object usually implies not much logic.

AHeise replied:
You can simply introduce a wrapper around state that also manages the producer. In that way, we don't need to think about the additional things in the serializer.

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaWriterState.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Just let IDE generate equals (and hashCode!)

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/Semantic.java:
```diff
@@ -0,0 +1,30 @@
+ * <li>{@link #AT_LEAST_ONCE}
+ * <li>{@link #NONE}
+ */
+public enum Semantic {
```
I know that this is the old name, but isn't this more ProcessingGuarantees? Should we factor that out to `flink-core`?

fapaul replied:
Why do you opt for `flink-core` and not something like `connector-base`? In general, I like the idea to move it somewhere reusable across multiple connectors.

AHeise replied:
`connector-base` are building blocks, `flink-core` is the core API. I'm not 100%, so maybe it's a good idea to start with `connector-base` and uplevel when needed.
Btw I also like to provide base sink/source builder where we could already implement the property.

fapaul replied:
I moved it to `connector base` for now and renamed it.

RollsBean committed:
[FLINK-23059][docs-zh] Translate update to operations playground docs to Chinese (#16279)

JingsongLi committed:
[FLINK-23142][table] UpdatableTopNFunction output wrong order in the same unique key

This closes #16284

zentol committed:
[FLINK-23196][tests] use MiniClusterResource

dianfu committed:
[FLINK-23213][python] Refactor the code strucuture of Python DataStream API

This PR refactors the following to make the code path simpler:
- Merge ProcessFunctionOperation and DataStreamStatelessFunctionOperation into StatelessFunctionOperation
- Move timerservice.py to package pyflink.fn_execution.datastream
- Merge OneInputRowWithTimerHandler and TwoInputRowWithTimerHandler into InputHandler

weizheng92 committed:
[FLINK-23200][docs][table] Fix code example typo in "Table API" page (#16338)

dawidwys committed:
[hotfix][formatting] Fix autoformatting which made javadoc of JobVertex#operatorIDs unreadable

tsreaper committed:
[FLINK-23149][table-code-splitter] Introduce the new Java code splitter module

This closes #16283

rovo98 committed:
[hotfix]: Fix broken javaDoc inline link.

dianfu committed:
[FLINK-23213][python] Move class SimpleStateRequestHandler into a separate file

dianfu committed:
[FLINK-23213][python] Refactor BeamPythonFunctionRunner to make it more readable by grouping similar methods together

dianfu committed:
[FLINK-23213][python] Remove BeamTableStatefulPythonFunctionRunner and BeamTableStatelessPythonFunctionRunner and merge them into BeamTablePythonFunctionRunner

KarmaGYZ committed:
[hotfix][core] Check the cpu and heap memory to be positive when building ResourceSpec

rookdorf committed:
[hotfix][docs] Fix variable name

leonardBang committed:
[hotfix][streaming] Fix the boundary condition when fetch limited records from streaming job

This closes #16256

AHeise commented on flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaProducerRecordSerializationSchema.java:
```diff
@@ -0,0 +1,64 @@
+     *
+     * @return topic id
+     */
+    String getTargetTopic(T record);
```
Why do we need this? The user would set the topic directly on the `ProducerRecord`...

dawidwys committed:
[FLINK-23095][runtime] Close the channel unspilling thread pool after restore

RollsBean committed:
[FLINK-23226][docs-zh] Fix Flink Chinese doc learn-flink/etl transformation.svg display issue

This closes #16364

dianfu committed:
[FLINK-23232][python] Use pickle.loads defined in pyflink to avoid race condition of the default pickle

tsreaper committed:
[FLINK-23215][table-code-splitter] Move NOTICE file into META-INF

This closes #16350

dawidwys committed:
[FLINK-22972][docs] Update description of a task lifecycle

hapihu committed:
[FLINK-23225][doc-zh] Fixed some translation mismatches in the "flink-architecture" page (#16363)

hapihu committed:
[FLINK-23260][docs] Fix the link for Page "docs/libs/gelly/overview" (#16382)

hapihu committed:
[FLINK-23259][docs] Fix the link for page "docs/dev/datastream/operators/overview" (#16381)
